### DISCUSSION TITLE
Support for gdb on m1?

### DISCUSSION THREAD

**Discussion by za-01** [Homebrew/discussions#1114] [Getting started]
I am trying to install gdb on MBA m1 with `brew install gdb`. It gives an error as
`Error: gdb: no bottle available!
You can try to install from source with:
  brew install --build-from-source gdb
Please note building from source is unsupported. You will encounter build
failures with some formulae. If you experience any issues please create pull
requests instead of asking for help on Homebrew's GitHub, Twitter or any other
official channels.`

Is there any other way to install gdb on an M1 machine?

EDIT: found [onlinegdb](https://www.onlinegdb.com), works good enough for now.

**Comment by carlocab** [comment]
Have you tried `brew install --build-from-source`?

If that doesn't work, it's likely that `gdb` just doesn't support M1. I want to say "doesn't support yet", but the `gdb` devs have always had a limited ability to support macOS, so I wouldn't hold my breath.

**Reply by za-01** [reply]
Yeah I did, didn't work :(

**Comment by nibin-matteo-git** [comment]
I too use a MBA M1
I simply downloaded the github desktop cuz I heard it comes along with all the necessary softwares and it works perfectly fine (i.e,u can use it in the terminal).

**Reply by za-01** [reply]
I have GitHub desktop installed but it doesn't work even in the terminal, did you do change some settings in GitHub desktop?

**Reply by nibin-matteo-git** [reply]
Not really.
I'm not sure what the problem is but reinstalling is worth a shot.

**Reply by ShashiKiranMR** [reply]
Is there any new solution for this problem?

**Reply by za-01** [reply]
Not afaik. But if you need C/C++ IDE with debugging on m1 install XCode. It supports debugging on C/C++, but it uses Clang, not gcc

**Reply by carlocab** [reply]
You can also use LLDB if you don't want/need the Xcode IDE.

**Comment by lyinch** [comment]
Hey, is there an update on this?

**Reply by gromgit** [reply]
Yes. GDB still doesn't support Apple Silicon, and there doesn't seem to be any indication on the mailing lists that anyone's even working on it.

I wouldn't hold my breath for this.

**Reply by SMillerDev** [reply]
The best place to ask is the gdb mailinglist. Homebrew only packages it.

**Comment by jiangshengdev** [comment]
**GDB Configuration for Apple Silicon in CLion.app**

It's been discovered that CLion comes with a version of GDB configured for Apple Silicon. To access it, navigate to:

```
cd /Users/{YOUR_USER_NAME}/Applications/CLion.app/Contents/bin/gdb/mac/aarch64/bin
```

Then run `./gdb --configuration` to see the configuration details. This version is configured with various options, including support for `aarch64-apple-darwin21.1` and is relocatable, meaning it can be moved within the GDB installation tree without issue.

```
./gdb --configuration
This GDB was configured as follows:
   configure --host=aarch64-apple-darwin21.1 --target=aarch64-apple-darwin21.1
	     --with-auto-load-dir=$debugdir:$datadir/auto-load
	     --with-auto-load-safe-path=$debugdir:$datadir/auto-load
	     --with-expat
	     --with-gdb-datadir=//share/gdb (relocatable)
	     --with-jit-reader-dir=//lib/gdb (relocatable)
	     --without-libunwind-ia64
	     --without-lzma
	     --without-babeltrace
	     --without-intel-pt
	     --with-mpfr
	     --without-xxhash
	     --without-python
	     --without-python-libdir
	     --without-debuginfod
	     --without-guile
	     --disable-source-highlight
	     --enable-threading
	     --with-separate-debug-dir=//lib/debug (relocatable)

("Relocatable" means the directory can be moved with the GDB installation
tree, and GDB will still find it.)

```

---

**Compiling GDB for RISC-V on Apple Silicon**

For those interested in debugging RISC-V platforms, here are the steps to compile a special version of GDB:

1. Download the gdb-13.2 source code.
2. Copy `isl 0.15` to the GDB source root directory, as per the prerequisites listed on [GCC's website](https://gcc.gnu.org/install/prerequisites.html). You can download `isl-0.15` directly from [here](https://gcc.gnu.org/pub/gcc/infrastructure/isl-0.15.tar.bz2).

Before compiling, configure your `.zshrc` file with the following exports to set up the necessary flags and paths:

```shell
export LDFLAGS="-L/opt/homebrew/opt/gmp/lib $LDFLAGS"
export CPPFLAGS="-I/opt/homebrew/opt/gmp/include $CPPFLAGS"
export LDFLAGS="-L/opt/homebrew/opt/libiconv/lib $LDFLAGS"
export CPPFLAGS="-I/opt/homebrew/opt/libiconv/include $CPPFLAGS"
export LDFLAGS="-L/opt/homebrew/opt/gettext/lib $LDFLAGS"
export CPPFLAGS="-I/opt/homebrew/opt/gettext/include $CPPFLAGS"
```

Then, in the gdb-13.2 source code root directory, execute:

```shell
mkdir -p build/ && cd build/
../configure --prefix=/usr/local --with-python=/opt/homebrew/Cellar/python@3.10/3.10.14/Frameworks/Python.framework/Versions/3.10/bin/python3.10 --target=riscv64-unknown-elf --enable-tui=yes --enable-werror=no
make -j$(sysctl -n hw.ncpu)
sudo make install
```

The result is a GDB configured for `riscv64-unknown-elf` with a similar set of options as the Apple Silicon version, also relocatable.

```
riscv64-unknown-elf-gdb --configuration

This GDB was configured as follows:
   configure --host=aarch64-apple-darwin23.4.0 --target=riscv64-unknown-elf
	     --with-auto-load-dir=:${prefix}/share/auto-load
	     --with-auto-load-safe-path=:${prefix}/share/auto-load
	     --with-expat
	     --with-gdb-datadir=/usr/local/share/gdb (relocatable)
	     --with-jit-reader-dir=/usr/local/lib/gdb (relocatable)
	     --without-libunwind-ia64
	     --without-lzma
	     --without-babeltrace
	     --without-intel-pt
	     --without-mpfr
	     --without-xxhash
	     --with-python=/opt/homebrew/opt/python@3.10/Frameworks/Python.framework/Versions/3.10
	     --with-python-libdir=/opt/homebrew/opt/python@3.10/Frameworks/Python.framework/Versions/3.10/lib
	     --without-debuginfod
	     --without-guile
	     --disable-source-highlight
	     --enable-threading
	     --with-separate-debug-dir=/usr/local/lib/debug (relocatable)

("Relocatable" means the directory can be moved with the GDB installation
tree, and GDB will still find it.)

```

---

This approach provides a viable solution for those needing GDB on Apple Silicon, particularly for specialized applications like RISC-V platform debugging.

**Comment by patryk4815** [comment]
will be fixed in https://github.com/Homebrew/homebrew-core/pull/209753

**Reply by gromgit** [reply]
With a blurb like that, not many folks would bother reading that PR, wherein you wrote:

> Still it is not possible to debug aarch64-darwin binaries via GDB.
> But GDB is useful as a frontend for remote debugging.

so folks wanting to replace `lldb` for macOS ARM development will be left disappointed.

**Reply by tomengland** [reply]
Hi All,

Can I ask how are people using gdb now it's installable?  Is it correct that it still doesn't work?

**Reply by patryk4815** [reply]
Yes it is installable using `brew install gdb`, but still you can't debug native Mach-O binaries. **GDB never worked good with Mach-O**. You should use `LLDB` for `Mach-O` debugging, and `GDB` for `ELF` remote debugging
